### PR TITLE
Week-4-libraries: Translation

### PR DESCRIPTION
**Pull request by rebekadittrich**

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -27,51 +30,76 @@
...
+    List<String> comments = new ArrayList<>();
     for (Entity entity : results.asIterable()) {
-      String greeting = (String) entity.getProperty("greeting");
-      greetings.add(greeting);
+      String comment = (String) entity.getProperty("comment");
+
+      String language = request.getParameter("languages");
+
+      if (language != null && !language.equals("original")) {
+        Translate translate = TranslateOptions.getDefaultInstance().getService();
```
Comment (marcbury-g) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: You probably only want to call this once and not each time you translate a message.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -27,51 +30,76 @@
...
+        Translation translation = translate.translate(comment, Translate.TranslateOption.targetLanguage(language));
```
Comment (marcbury-g) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: It is better to reduce the number of calls to translation API for better latency. There is a [translate function ](https://googleapis.dev/java/google-cloud-translate/latest/com/google/cloud/translate/Translate.html#translate-java.util.List-com.google.cloud.translate.Translate.TranslateOption...-) that also accepts a list of strings.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 3]
```diff
@@ -27,51 +30,76 @@
...
     PreparedQuery results = datastore.prepare(query);
 
-    List<String> greetings = new ArrayList<>();
+    List<String> comments = new ArrayList<>();
     for (Entity entity : results.asIterable()) {
-      String greeting = (String) entity.getProperty("greeting");
-      greetings.add(greeting);
+      String comment = (String) entity.getProperty("comment");
+
+      String language = request.getParameter("languages");
```
Comment (marcbury-g) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Why is the param called languages (plural)?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -27,51 +30,76 @@
...
-  private String convertGreetingsToJson(List<String> greetings) {
+  /* Convert comments ArrayList into Json. */
+  private String convertCommentsToJson(List<String> comments) {
     
-    int length = greetings.size();
+    int length = comments.size();
 
     String json = "[";
     for (int i = 0; i < length; i++) {
-      json += "{\"greeting\":\"";
```
Comment (marcbury-g) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:73]: Why do you not use Gson? https://github.com/google/gson/blob/master/UserGuide.md#primitives-examples
Either you make a new class for the Comment or you just return a list of strings.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -27,51 +30,76 @@
...
+        Translation translation = translate.translate(comment, Translate.TranslateOption.targetLanguage(language));
+        comment = translation.getTranslatedText();
+      }
+
+      comments.add(comment);
     }
     
-    response.setContentType("text/html;");
-    response.getWriter().println(convertGreetingsToJson(greetings));
+    response.setContentType("text/html");
```
Comment (marcbury-g) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Why do you have content type html if you return json? You can use application/json

File: portfolio/src/main/webapp/script.js [hunk 6]
```diff
@@ -12,30 +12,32 @@
...
 
-  // Add it to the page.
-  const greetingContainer = document.getElementById('greeting-container');
-  greetingContainer.innerText = greeting;
+/* Add all fetched comments to the page. */
+function getComments(language) {
+  document.getElementById('comment-container').innerText = "";
+  fetch('/data?' + new URLSearchParams({
+    languages: language,
+})).then(response => response.json()).then(comment => comment.forEach(addComment));
```
Comment (marcbury-g) [inline portfolio/src/main/webapp/script.js]: Please indicate in the name that you expect that there is a list returned: comments => comments.forEach...

**Review by marcbury-g — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 2 above)
Comment (rebekadittrich) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: I couldn't find a getTranslatedText() method for lists so I would have to iterate through the list twice manually if I want to add the elements, translate them, then add the translated back to the list of strings. Wouldn't this make the process longer than the latency in the original case?

File: portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java [hunk 7]
```diff
@@ -19,65 +19,83 @@
...
 import java.util.ArrayList;
 import java.util.List;
+import java.util.stream.*;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
+
 /** Servlet that returns some example content. */
```
Comment (pmueller-gg) [inline portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java]: The comment seems not to be up-to-date with this change.

File: portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java [hunk 8]
```diff
@@ -19,65 +19,83 @@
...
 @WebServlet("/data")
-public class DataServlet extends HttpServlet {
+public class CommentServlet extends HttpServlet {
```
Comment (pmueller-gg) [inline portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java:38]:
```
@WebServlet("/data")
public class CommentServlet
```
Should this be "/comment"? Or is the naming on purpose?

File: portfolio/src/main/webapp/script.js [hunk 9]
```diff
@@ -12,30 +12,32 @@
...
+  const safeComment = safeEncoding(comment.comment);
+  document.getElementById('comment-container').innerText += safeComment + "\n";
 }
 
-/* Add value of greeting field to the page. */
-function addGreeting(greeting) {
-  const safeGreeting = greeting.greeting.replace(/</g, "&lt;").replace(/>/g, "&gt;");
-  document.getElementById('greeting-container').innerText += safeGreeting + "\n";
+/* Replace symbols in text for safety. */
+function safeEncoding(text) {
```
Comment (pmueller-gg) [inline portfolio/src/main/webapp/script.js:41]: Does this kind of cleanup of user input happen server-side as well?

File: portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java [hunk 10]
```diff
@@ -19,65 +19,83 @@
...
+    Translate translate = TranslateOptions.getDefaultInstance().getService();
+
+    List<Comment> comments = new ArrayList<>();
     for (Entity entity : results.asIterable()) {
-      String greeting = (String) entity.getProperty("greeting");
-      greetings.add(greeting);
+      String text = (String) entity.getProperty("comment");
+
+      if (isNotOriginal) {
+        Translation translation = translate.translate(text, Translate.TranslateOption.targetLanguage(language));
```
Comment (pmueller-gg) [inline portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java:71]: How do you guarantee that the target language is supported?

**Review by pmueller-gg — APPROVED** [review]

**Review by marcbury-g — APPROVED** [review]

**Pull request merged**